### PR TITLE
Add basic telemetry to collector

### PR DESCRIPTION
**Pull request by pjanotti**
This change adds a Prometheus endpoint using OC to provide metrics
to the collector. There are settings to control the level and port
in which the metrics are exposed. The main metrics added for now
are in the queued exporters.

A0-573

### CODE REVIEW

**Comment by pjanotti** [comment]
Example of metrics in most verbose mode:
```
# HELP oc_collector_batches The number of span batches received.
# TYPE oc_collector_batches counter
oc_collector_batches{exporter="jaeger-error",format="oc",service="fstSvc"} 18
oc_collector_batches{exporter="jaeger-error",format="oc",service="sndSvc"} 18
oc_collector_batches{exporter="jaeger-error",format="oc",service="trdSvc"} 22
oc_collector_batches{exporter="jaeger-ok",format="oc",service="fstSvc"} 18
oc_collector_batches{exporter="jaeger-ok",format="oc",service="sndSvc"} 18
oc_collector_batches{exporter="jaeger-ok",format="oc",service="trdSvc"} 22
# HELP oc_collector_dropped_batches The number of span batches dropped.
# TYPE oc_collector_dropped_batches counter
oc_collector_dropped_batches{exporter="jaeger-error",format="oc",service="fstSvc"} 3
oc_collector_dropped_batches{exporter="jaeger-error",format="oc",service="sndSvc"} 3
oc_collector_dropped_batches{exporter="jaeger-error",format="oc",service="trdSvc"} 2
# HELP oc_collector_dropped_spans The number of spans dropped.
# TYPE oc_collector_dropped_spans untyped
oc_collector_dropped_spans{exporter="jaeger-error",format="oc",service="fstSvc"} 607
oc_collector_dropped_spans{exporter="jaeger-error",format="oc",service="sndSvc"} 571
oc_collector_dropped_spans{exporter="jaeger-error",format="oc",service="trdSvc"} 406
# HELP oc_collector_fail_send The number of failed send operations performed by queued exporter
# TYPE oc_collector_fail_send counter
oc_collector_fail_send{exporter="jaeger-error",format="oc",service="fstSvc"} 7
oc_collector_fail_send{exporter="jaeger-error",format="oc",service="sndSvc"} 8
oc_collector_fail_send{exporter="jaeger-error",format="oc",service="trdSvc"} 30
# HELP oc_collector_queue_length Current number of batches in the queued exporter
# TYPE oc_collector_queue_length gauge
oc_collector_queue_length{exporter="jaeger-error"} 50
oc_collector_queue_length{exporter="jaeger-ok"} 0
# HELP oc_collector_send_latency The latency of the successful send operations.
# TYPE oc_collector_send_latency histogram
oc_collector_send_latency_bucket{exporter="jaeger-ok",format="oc",service="fstSvc",le="10"} 10
oc_collector_send_latency_bucket{exporter="jaeger-ok",format="oc",service="fstSvc",le="25"} 18
oc_collector_send_latency_bucket{exporter="jaeger-ok",format="oc",service="fstSvc",le="50"} 18
oc_collector_send_latency_bucket{exporter="jaeger-ok",format="oc",service="fstSvc",le="75"} 18
oc_collector_send_latency_bucket{exporter="jaeger-ok",format="oc",service="fstSvc",le="100"} 18
oc_collector_send_latency_bucket{exporter="jaeger-ok",format="oc",service="fstSvc",le="250"} 18
oc_collector_send_latency_bucket{exporter="jaeger-ok",format="oc",service="fstSvc",le="500"} 18
oc_collector_send_latency_bucket{exporter="jaeger-ok",format="oc",service="fstSvc",le="750"} 18
oc_collector_send_latency_bucket{exporter="jaeger-ok",format="oc",service="fstSvc",le="1000"} 18
oc_collector_send_latency_bucket{exporter="jaeger-ok",format="oc",service="fstSvc",le="2000"} 18
oc_collector_send_latency_bucket{exporter="jaeger-ok",format="oc",service="fstSvc",le="3000"} 18
oc_collector_send_latency_bucket{exporter="jaeger-ok",format="oc",service="fstSvc",le="4000"} 18
oc_collector_send_latency_bucket{exporter="jaeger-ok",format="oc",service="fstSvc",le="5000"} 18
oc_collector_send_latency_bucket{exporter="jaeger-ok",format="oc",service="fstSvc",le="10000"} 18
oc_collector_send_latency_bucket{exporter="jaeger-ok",format="oc",service="fstSvc",le="20000"} 18
oc_collector_send_latency_bucket{exporter="jaeger-ok",format="oc",service="fstSvc",le="30000"} 18
oc_collector_send_latency_bucket{exporter="jaeger-ok",format="oc",service="fstSvc",le="50000"} 18
oc_collector_send_latency_bucket{exporter="jaeger-ok",format="oc",service="fstSvc",le="+Inf"} 18
oc_collector_send_latency_sum{exporter="jaeger-ok",format="oc",service="fstSvc"} 172.99999999999997
oc_collector_send_latency_count{exporter="jaeger-ok",format="oc",service="fstSvc"} 18
oc_collector_send_latency_bucket{exporter="jaeger-ok",format="oc",service="sndSvc",le="10"} 15
oc_collector_send_latency_bucket{exporter="jaeger-ok",format="oc",service="sndSvc",le="25"} 18
oc_collector_send_latency_bucket{exporter="jaeger-ok",format="oc",service="sndSvc",le="50"} 18
oc_collector_send_latency_bucket{exporter="jaeger-ok",format="oc",service="sndSvc",le="75"} 18
oc_collector_send_latency_bucket{exporter="jaeger-ok",format="oc",service="sndSvc",le="100"} 18
oc_collector_send_latency_bucket{exporter="jaeger-ok",format="oc",service="sndSvc",le="250"} 18
oc_collector_send_latency_bucket{exporter="jaeger-ok",format="oc",service="sndSvc",le="500"} 18
oc_collector_send_latency_bucket{exporter="jaeger-ok",format="oc",service="sndSvc",le="750"} 18
oc_collector_send_latency_bucket{exporter="jaeger-ok",format="oc",service="sndSvc",le="1000"} 18
oc_collector_send_latency_bucket{exporter="jaeger-ok",format="oc",service="sndSvc",le="2000"} 18
oc_collector_send_latency_bucket{exporter="jaeger-ok",format="oc",service="sndSvc",le="3000"} 18
oc_collector_send_latency_bucket{exporter="jaeger-ok",format="oc",service="sndSvc",le="4000"} 18
oc_collector_send_latency_bucket{exporter="jaeger-ok",format="oc",service="sndSvc",le="5000"} 18
oc_collector_send_latency_bucket{exporter="jaeger-ok",format="oc",service="sndSvc",le="10000"} 18
oc_collector_send_latency_bucket{exporter="jaeger-ok",format="oc",service="sndSvc",le="20000"} 18
oc_collector_send_latency_bucket{exporter="jaeger-ok",format="oc",service="sndSvc",le="30000"} 18
oc_collector_send_latency_bucket{exporter="jaeger-ok",format="oc",service="sndSvc",le="50000"} 18
oc_collector_send_latency_bucket{exporter="jaeger-ok",format="oc",service="sndSvc",le="+Inf"} 18
oc_collector_send_latency_sum{exporter="jaeger-ok",format="oc",service="sndSvc"} 147.00000000000003
oc_collector_send_latency_count{exporter="jaeger-ok",format="oc",service="sndSvc"} 18
oc_collector_send_latency_bucket{exporter="jaeger-ok",format="oc",service="trdSvc",le="10"} 17
oc_collector_send_latency_bucket{exporter="jaeger-ok",format="oc",service="trdSvc",le="25"} 22
oc_collector_send_latency_bucket{exporter="jaeger-ok",format="oc",service="trdSvc",le="50"} 22
oc_collector_send_latency_bucket{exporter="jaeger-ok",format="oc",service="trdSvc",le="75"} 22
oc_collector_send_latency_bucket{exporter="jaeger-ok",format="oc",service="trdSvc",le="100"} 22
oc_collector_send_latency_bucket{exporter="jaeger-ok",format="oc",service="trdSvc",le="250"} 22
oc_collector_send_latency_bucket{exporter="jaeger-ok",format="oc",service="trdSvc",le="500"} 22
oc_collector_send_latency_bucket{exporter="jaeger-ok",format="oc",service="trdSvc",le="750"} 22
oc_collector_send_latency_bucket{exporter="jaeger-ok",format="oc",service="trdSvc",le="1000"} 22
oc_collector_send_latency_bucket{exporter="jaeger-ok",format="oc",service="trdSvc",le="2000"} 22
oc_collector_send_latency_bucket{exporter="jaeger-ok",format="oc",service="trdSvc",le="3000"} 22
oc_collector_send_latency_bucket{exporter="jaeger-ok",format="oc",service="trdSvc",le="4000"} 22
oc_collector_send_latency_bucket{exporter="jaeger-ok",format="oc",service="trdSvc",le="5000"} 22
oc_collector_send_latency_bucket{exporter="jaeger-ok",format="oc",service="trdSvc",le="10000"} 22
oc_collector_send_latency_bucket{exporter="jaeger-ok",format="oc",service="trdSvc",le="20000"} 22
oc_collector_send_latency_bucket{exporter="jaeger-ok",format="oc",service="trdSvc",le="30000"} 22
oc_collector_send_latency_bucket{exporter="jaeger-ok",format="oc",service="trdSvc",le="50000"} 22
oc_collector_send_latency_bucket{exporter="jaeger-ok",format="oc",service="trdSvc",le="+Inf"} 22
oc_collector_send_latency_sum{exporter="jaeger-ok",format="oc",service="trdSvc"} 203.00000000000003
oc_collector_send_latency_count{exporter="jaeger-ok",format="oc",service="trdSvc"} 22
# HELP oc_collector_spans The number of spans received.
# TYPE oc_collector_spans untyped
oc_collector_spans{exporter="jaeger-error",format="oc",service="fstSvc"} 3478
oc_collector_spans{exporter="jaeger-error",format="oc",service="sndSvc"} 3497
oc_collector_spans{exporter="jaeger-error",format="oc",service="trdSvc"} 4280
oc_collector_spans{exporter="jaeger-ok",format="oc",service="fstSvc"} 3478
oc_collector_spans{exporter="jaeger-ok",format="oc",service="sndSvc"} 3497
oc_collector_spans{exporter="jaeger-ok",format="oc",service="trdSvc"} 4280
# HELP oc_collector_success_send The number of successful send operations performed by queued exporter
# TYPE oc_collector_success_send counter
oc_collector_success_send{exporter="jaeger-ok",format="oc",service="fstSvc"} 18
oc_collector_success_send{exporter="jaeger-ok",format="oc",service="sndSvc"} 18
oc_collector_success_send{exporter="jaeger-ok",format="oc",service="trdSvc"} 22
```

File: internal/collector/processor/queued_processor.go [hunk 1]
```diff
@@ -54,13 +60,30 @@ func NewQueuedSpanProcessor(sender SpanProcessor, opts ...Option) SpanProcessor
 		sp.processItemFromQueue(value)
 	})
 
+	// Start a timer to report the queue length.
+	ticker := time.NewTicker(1 * time.Second)
+	go func() {
+		defer ticker.Stop()
+		ctx, _ := tag.New(context.Background(), tag.Upsert(tagProcessorNameKey, sp.name))
```
Comment (maynk) [inline internal/collector/processor/queued_processor.go]: perhaps context should be passed in, instead of using context.Background? This could allow more "inherited" tags down the line

File: internal/collector/processor/queued_processor.go [hunk 2]
```diff
@@ -94,50 +117,149 @@ func (sp *queuedSpanProcessor) enqueueSpanBatch(batch *agenttracepb.ExportTraceS
...
+	numSpans := len(item.batch.Spans)
+	stats.RecordWithTags(context.Background(), statsTags, statDroppedBatchCount.M(1), statDroppedSpanCount.M(int64(numSpans)))
+
 	sp.logger.Warn("Span batch dropped",
+		zap.String("processor", sp.name),
 		zap.Int("#spans", len(item.batch.Spans)),
 		zap.String("spanSource", item.spanFormat))
 }
+
+// Variables related to metrics specific to queued processor.
```
Comment (maynk) [inline internal/collector/processor/queued_processor.go]: is there a convention? e.g. to put all variable declarations at top of file?

File: internal/collector/processor/queued_processor.go [hunk 3]
```diff
@@ -94,50 +117,149 @@ func (sp *queuedSpanProcessor) enqueueSpanBatch(batch *agenttracepb.ExportTraceS
...
+	tagKeys := metricTagKeys(level)
+	if tagKeys == nil {
+		return nil
+	}
+
+	queueLengthView := &view.View{
+		Name:        statQueueLength.Name(),
+		Measure:     statQueueLength,
+		Description: "Current number of batches in the queued exporter",
+		TagKeys:     []tag.Key{tagProcessorNameKey},
```
Comment (maynk) [inline internal/collector/processor/queued_processor.go]: why is this not tagKeys like the views below?

File: internal/collector/processor/queued_processor.go [hunk 4]
```diff
@@ -94,50 +117,149 @@ func (sp *queuedSpanProcessor) enqueueSpanBatch(batch *agenttracepb.ExportTraceS
...
+	}
+
+	var latencyDistributionAggregation *view.Aggregation
+	switch level {
+	case telemetry.Basic:
+		latencyDistributionAggregation = view.Distribution(100, 500, 1000, 5000)
+	case telemetry.Normal:
+		latencyDistributionAggregation = view.Distribution(50, 100, 250, 500, 1000, 2500, 5000, 10000)
+	case telemetry.Detailed:
+		latencyDistributionAggregation = view.Distribution(10, 25, 50, 75, 100, 250, 500, 750, 1000, 2000, 3000, 4000, 5000, 10000, 20000, 30000, 50000)
```
Comment (maynk) [inline internal/collector/processor/queued_processor.go]: why not always keep the detailed distribution? is it really that expensive?

File: internal/collector/processor/queued_processor.go (hunk 2 above)
Comment (pjanotti) [reply internal/collector/processor/queued_processor.go]: I actually don't know: my idea was to keep it separate from other concerns, but let me look at some other golang code to see how different projects deal with this.

File: internal/collector/processor/queued_processor.go (hunk 3 above)
Comment (pjanotti) [reply internal/collector/processor/queued_processor.go]: The queue contains entries from different "format" and "service" so the length only applies to the name of the queue.

File: internal/collector/processor/queued_processor.go (hunk 4 above)
Comment (pjanotti) [reply internal/collector/processor/queued_processor.go]: Cardinality can go high: service X format X distribution. That said I think we have to start at some point and let customers give their feedback so we can tune those. Ideally we should have configs to allow flexibility on tags and distribution but I didn't want to add that right now to keep the change small.

File: cmd/occollector/app/collector/collector.go [hunk 5]
```diff
@@ -176,6 +209,49 @@ func execute() {
...
+			if serveErr != nil && serveErr != http.ErrServerClosed {
+				logger.Error("Failed to serve metrics", zap.Error(serveErr))
+			}
+		default:
+		}
+	}()
+
+	select {
+	case err = <-errChan:
+	case <-time.After(500*time.Millisecond):
```
Comment (maynk) [inline cmd/occollector/app/collector/collector.go]: so effectively initTelemetry blocks startup for 500ms? is there an alternative?

File: internal/collector/processor/processor.go [hunk 6]
```diff
@@ -37,11 +43,109 @@ func (sp *debugSpanProcessor) ProcessSpans(batch *agenttracepb.ExportTraceServic
...
+// NewDebugSpanProcessor creates an OC SpanProcessor that just drops the received data.
+func NewDebugSpanProcessor(logger *zap.Logger) SpanProcessor {
 	return &debugSpanProcessor{logger: logger}
 }
+
+// Keys and stats for telemetry.
+var (
+	tagSourceFormatKey, _  = tag.NewKey("format")
+	tagServiceNameKey, _   = tag.NewKey("service")
+	tagProcessorNameKey, _ = tag.NewKey("exporter")
```
Comment (maynk) [inline internal/collector/processor/processor.go]: should it not be called tagExporterNameKey?

File: internal/collector/processor/processor.go [hunk 7]
```diff
@@ -37,11 +43,109 @@ func (sp *debugSpanProcessor) ProcessSpans(batch *agenttracepb.ExportTraceServic
...
+
+	statBatchCount        = stats.Int64("batches", "counts the number of batches received", stats.UnitDimensionless)
```
Comment (maynk) [inline internal/collector/processor/processor.go]: "batches" -> "received_batches"?

File: internal/collector/processor/processor.go [hunk 8]
```diff
@@ -37,11 +43,109 @@ func (sp *debugSpanProcessor) ProcessSpans(batch *agenttracepb.ExportTraceServic
...
+	statDroppedBatchCount = stats.Int64("dropped_batches", "counts the number of batches dropped", stats.UnitDimensionless)
+
+	statSpanCount        = stats.Int64("spans", "counts the number of spans received", stats.UnitDimensionless)
```
Comment (maynk) [inline internal/collector/processor/processor.go]: "spans" -> "received_spans"?

File: internal/collector/processor/processor.go [hunk 9]
```diff
@@ -37,11 +43,109 @@ func (sp *debugSpanProcessor) ProcessSpans(batch *agenttracepb.ExportTraceServic
...
+	numSpans := len(batch.Spans)
+	stats.RecordWithTags(context.Background(), statsTags, statBatchCount.M(1), statSpanCount.M(int64(numSpans)))
+
+	sp.logger.Debug("debugSpanProcessor", zap.String("originalFormat", spanFormat), zap.Int("#spans", numSpans))
 	return 0, nil
 }
 
-// NewNoopSpanProcessor creates an OC SpanProcessor that just drops the received data.
-func NewNoopSpanProcessor(logger *zap.Logger) SpanProcessor {
+// NewDebugSpanProcessor creates an OC SpanProcessor that just drops the received data.
```
Comment (maynk) [inline internal/collector/processor/processor.go]: why did you rename NoopSP to DebugSP - because now it provides telemetry? IMO Noop was explaining the functionality better to me, but it's just my opinion.

File: internal/collector/processor/processor.go [hunk 10]
```diff
@@ -37,11 +43,109 @@ func (sp *debugSpanProcessor) ProcessSpans(batch *agenttracepb.ExportTraceServic
...
+
+// MetricViews return the metrics views according to given telemetry level.
+func MetricViews(level telemetry.Level) []*view.View {
+	tagKeys := metricTagKeys(level)
+	if tagKeys == nil {
+		return nil
+	}
+
+	// There are some metrics enabled, return the views.
+	countBatchesView := &view.View{
```
Comment (maynk) [inline internal/collector/processor/processor.go]: receivedBatchesView?

File: internal/collector/processor/processor.go [hunk 11]
```diff
@@ -37,11 +43,109 @@ func (sp *debugSpanProcessor) ProcessSpans(batch *agenttracepb.ExportTraceServic
...
+		Aggregation: view.Count(),
+	}
+	droppedBatchesView := &view.View{
+		Name:        statDroppedBatchCount.Name(),
+		Measure:     statDroppedBatchCount,
+		Description: "The number of span batches dropped.",
+		TagKeys:     tagKeys,
+		Aggregation: view.Count(),
+	}
+	countSpansView := &view.View{
```
Comment (maynk) [inline internal/collector/processor/processor.go]: receivedSpansView?

File: cmd/occollector/app/collector/collector.go [hunk 12]
```diff
@@ -136,18 +156,22 @@ func execute() {
...
-	for _, queuedJaegerProcessorCfg := range multiProcessorCfg.Processors {
+	for i, queuedJaegerProcessorCfg := range multiProcessorCfg.Processors {
 		logger.Info("Queued Jaeger Sender Enabled")
 		queuedJaegerProcessor, err := buildQueuedSpanProcessor(logger, queuedJaegerProcessorCfg)
 		if err != nil {
 			logger.Error("Failed to build the queued span processor", zap.Error(err))
 			os.Exit(1)
 		}
 		spanProcessors = append(spanProcessors, queuedJaegerProcessor)
+		if i == 0 {
```
[the remark this answers is not in the document]
Comment (pjanotti) [reply cmd/occollector/app/collector/collector.go]: Yeah, it will require a bool but likely will look cleaner.

[the remark this answers is not in the document]
Comment (maynk) [reply cmd/occollector/app/collector/collector.go]: what is the bool required?

File: internal/collector/processor/queued_processor.go (hunk 1 above)
Comment (pjanotti) [reply internal/collector/processor/queued_processor.go]: I will delay that until we have a case for that, somehow I feel strange passing the context to a New* function.

File: internal/collector/processor/queued_processor.go (hunk 4 above)
Comment (pjanotti) [reply internal/collector/processor/queued_processor.go]: Per conversation removed some of the tags for latency and kept only largest distribution.

File: cmd/occollector/app/collector/collector.go (hunk 12 above)
[the remark this answers is not in the document]
Comment (pjanotti) [reply cmd/occollector/app/collector/collector.go]: I was being "too clear" and adding the queued processor only views only if one had being created, but I don't see a bad side effect of adding it even if not being used. We can always change later if we discover a problem with that.

File: cmd/occollector/app/collector/collector.go (hunk 5 above)
Comment (pjanotti) [reply cmd/occollector/app/collector/collector.go]: Per conversation switched to a channel for errors, nice idea!

File: cmd/occollector/app/collector/collector.go [hunk 13]
```diff
@@ -176,6 +209,49 @@ func execute() {
...
+		Namespace: "oc_collector",
+	}
+	pe, err := prometheus.NewExporter(opts)
+	if err != nil {
+		return err
+	}
+
+	view.RegisterExporter(pe)
+
+	errChan := make(chan error, 1)
```
[the remark this answers is not in the document]
Comment (pjanotti) [reply cmd/occollector/app/collector/collector.go]: Using error chan per conversation

File: internal/collector/processor/processor.go (hunk 9 above)
Comment (pjanotti) [reply internal/collector/processor/processor.go]: In the command line it is called --debug-processor but let's keep as it is, avoiding opportunistic change.

File: internal/collector/processor/processor.go (hunk 11 above)
Comment (pjanotti) [reply internal/collector/processor/processor.go]: done
